src/builder/commands/copy.rs

```diff
@@ -43,32 +90,31 @@ impl BuildStep for Copy {
         if src.starts_with("/work") {
             match symlink_metadata(src) {
                 Ok(ref meta) if meta.file_type().is_dir() => {
-                    let re = try!(Regex::new(&self.ignore_regex)
+                    let ignore_re = try!(self.compile_ignore_regex()
+                        .map_err(|e| VersionError::Regex(Box::new(e))));
+                    let include_re = try!(self.compile_include_regex()
                         .map_err(|e| VersionError::Regex(Box::new(e))));
                     try!(ScanDir::all().walk(src, |iter| {
-                        let mut all_entries = iter.filter_map(|(e, _)|
-                        {
-                            let fpath = e.path();
-                            let strpath = {
+                        let all_paths: BTreeSet<_> = iter
+                            .filter_map(|(entry, _)| {
+                                let fpath = entry.path();
                                 // We know that directory is inside
                                 // the source
                                 let path = fpath.strip_prefix(src).unwrap();
-                                // We know that it's decodable
+                                // We know that path is decodable
                                 let strpath = path.to_str().unwrap();
-                                if !re.is_match(strpath) {
-                                    Some(strpath.to_string())
-                                } else {
-                                    None
-                                }
-                            };
-                            strpath.map(|x| (fpath, x))
-                        }).collect::<Vec<_>>();
-                        all_entries.sort();
-                        for (fpath, name) in all_entries {
-                            hash.field("filename", name);
-                            try!(hash.file(&fpath,
-                                self.owner_uid, self.owner_gid)
-                                .map_err(|e| VersionError::Io(e, fpath)));
+                                generate_paths(strpath,
+                                    ignore_re.as_ref(), include_re.as_ref())
```

Comment:
You can return empty vector here instead of None, and use flat_map directly.
But I'm not sure if returning many one-element vectors is a good idea. Maybe there is another way that is also simple enough?